### PR TITLE
feat: reference Secrets in `envs` and `volumes` sections in config

### PR DESCRIPTION
**Pull request by zroubalik**
Signed-off-by: Zbynek Roubalik <zroubali@redhat.com>

Users can reference Kubernetes Secrets in the `func.yaml`, in `envs` and `volumes` sections, see the example:

```yaml
name: test
namespace: ""
runtime: go
...
volumes:
- secret: mysecret3
  path: /workspace/secret
envs:
- name: EXAMPLE1
  value: myvalue
- name: EXAMPLE2
  value: '{{ env.LOCAL_ENV_VALUE }}'
- name: EXAMPLE3
  value: '{{ secret.mysecret2.EXAMPLE3 }}'
- value: '{{ secret.mysecret }}'
```

If some Secret is not present in the cluster, function is not being deployed:
```bash
$ ./func deploy -p test                                                                                                                                           
🕒 Pushing function image to the registry
🕖 Deploying function to the cluster
Error: knative deployer failed to update the Knative Service: 
  referenced Secret "mysecret" is not present in namespace "default"
  referenced Secret "mysecret2" is not present in namespace "default"
  referenced Secret "mysecret3" is not present in namespace "default"
```

If the entry in `func.yaml` is not valid, function is not being deployed:
```bash
./func deploy -p test                                                                                                                                                 
Error: 'func.yaml' config file is not valid:
  env entry #2 has invalid value field set, it has '{{ secretx.mysecret }}', but allowed is only '{{ secret.secretName }}'
  env entry #3 with name 'EXAMPLE3' has invalid value field set, it has '{{ secretx.mysecret2.EXAMPLE3 }}', but allowed is only '{{ env.MY_ENV }}' or '{{ secret.secretName.key }}'
```

Some parts of the code are prepared to refer Kubernetes ConfigMaps as well. A specific `func` command and interactive wizard will be done in follow up PR.

Relates: #362

### CODE REVIEW

**Comment by zroubalik** [comment]
PR https://github.com/boson-project/func/pull/368 needs to be merged first in order to rebuild & merge this one.

**Review by lkingland — APPROVED** [review]
What a huge addition; very nice!   Looks good to me; especially the solid test coverage

File: knative/client.go [hunk 1]
```diff
@@ -49,6 +50,16 @@ func NewEventingClient(namespace string) (clienteventingv1beta1.KnEventingClient
 	return client, nil
 }
 
+func NewKubernetesClientset(namespace string) (*kubernetes.Clientset, error) {
```
Comment (lance) [inline knative/client.go:53]: This function is not referenced anywhere else in the PR. Is it intended? Maybe I'm missing something.

**Review by lance — COMMENTED** [review]
Very nice! Just a few minor comments/suggestions and a question.

File: knative/client.go (hunk 1 above)
Comment (zroubalik) [reply knative/client.go:53]: It is being used in the function that checks whether Secret is present in the cluster:
https://github.com/boson-project/func/pull/369/files#diff-3bc17fd6241d546478336fa19a33f7e7f4c232421ce4e8511512c00fa617860dR490

**Pull request merged**